### ISSUE TITLE
java.lang.IndexOutOfBoundsException

### ISSUE THREAD

**Issue by Hipapheralkus** [DanNegrea/TokenJar#2]
Hi,
I noticed that when I enable and run Tokenjar, I get many Errors I'm using Burp Pro v1.7.33 64-bit.

```
	at burp.BurpExtender.processHttpMessage(BurpExtender.java:78)
	at burp.vec.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
java.lang.IndexOutOfBoundsException: Index: 6, Size: 4
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at tokenJar.DataModel.getByName(DataModel.java:136)
	at burp.BurpExtender.processRequestMessage(BurpExtender.java:184)
	at burp.BurpExtender.processHttpMessage(BurpExtender.java:78)
	at burp.vec.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
java.lang.IndexOutOfBoundsException: Index: 6, Size: 4
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at tokenJar.DataModel.getByName(DataModel.java:136)
	at burp.BurpExtender.processRequestMessage(BurpExtender.java:184)
	at burp.BurpExtender.processHttpMessage(BurpExtender.java:78)
	at burp.vec.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
java.lang.IndexOutOfBoundsException: Index: 6, Size: 4
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at tokenJar.DataModel.getByName(DataModel.java:136)
	at burp.BurpExtender.processRequestMessage(BurpExtender.java:184)
	at burp.BurpExtender.processHttpMessage(BurpExtender.java:78)
	at burp.vec.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
java.lang.IndexOutOfBoundsException: Index: 6, Size: 4
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at tokenJar.DataModel.getByName(DataModel.java:136)
	at burp.BurpExtender.processRequestMessage(BurpExtender.java:184)
	at burp.BurpExtender.processHttpMessage(BurpExtender.java:78)
	at burp.vec.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
java.lang.IndexOutOfBoundsException: Index: 6, Size: 4
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at tokenJar.DataModel.getByName(DataModel.java:136)
	at burp.BurpExtender.processRequestMessage(BurpExtender.java:184)
	at burp.BurpExtender.processHttpMessage(BurpExtender.java:78)
	at burp.vec.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
java.lang.IndexOutOfBoundsException: Index: 6, Size: 4
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at tokenJar.DataModel.getByName(DataModel.java:136)
	at burp.BurpExtender.processRequestMessage(BurpExtender.java:184)
	at burp.BurpExtender.processHttpMessage(BurpExtender.java:78)
	at burp.vec.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
java.lang.IndexOutOfBoundsException: Index: 6, Size: 4
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at tokenJar.DataModel.getByName(DataModel.java:136)
	at burp.BurpExtender.processRequestMessage(BurpExtender.java:184)
	at burp.BurpExtender.processHttpMessage(BurpExtender.java:78)
	at burp.vec.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
java.lang.IndexOutOfBoundsException: Index: 6, Size: 4
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at tokenJar.DataModel.getByName(DataModel.java:136)
	at burp.BurpExtender.processRequestMessage(BurpExtender.java:184)
	at burp.BurpExtender.processHttpMessage(BurpExtender.java:78)
	at burp.vec.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
java.lang.IndexOutOfBoundsException: Index: 6, Size: 4
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at tokenJar.DataModel.getByName(DataModel.java:136)
	at burp.BurpExtender.processRequestMessage(BurpExtender.java:184)
	at burp.BurpExtender.processHttpMessage(BurpExtender.java:78)
	at burp.vec.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
java.lang.IndexOutOfBoundsException: Index: 6, Size: 4
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at tokenJar.DataModel.getByName(DataModel.java:136)
	at burp.BurpExtender.processRequestMessage(BurpExtender.java:184)
	at burp.BurpExtender.processHttpMessage(BurpExtender.java:78)
	at burp.vec.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
java.lang.IndexOutOfBoundsException: Index: 6, Size: 4
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at tokenJar.DataModel.getByName(DataModel.java:136)
	at burp.BurpExtender.processRequestMessage(BurpExtender.java:184)
	at burp.BurpExtender.processHttpMessage(BurpExtender.java:78)
	at burp.vec.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
java.lang.IndexOutOfBoundsException: Index: 6, Size: 4
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at tokenJar.DataModel.getByName(DataModel.java:136)
	at burp.BurpExtender.processRequestMessage(BurpExtender.java:184)
	at burp.BurpExtender.processHttpMessage(BurpExtender.java:78)
	at burp.vec.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
java.lang.IndexOutOfBoundsException: Index: 6, Size: 4
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at tokenJar.DataModel.getByName(DataModel.java:136)
	at burp.BurpExtender.processRequestMessage(BurpExtender.java:184)
	at burp.BurpExtender.processHttpMessage(BurpExtender.java:78)
	at burp.vec.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
java.lang.IndexOutOfBoundsException: Index: 6, Size: 4
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at tokenJar.DataModel.getByName(DataModel.java:136)
	at burp.BurpExtender.processRequestMessage(BurpExtender.java:184)
	at burp.BurpExtender.processHttpMessage(BurpExtender.java:78)
	at burp.vec.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
java.lang.IndexOutOfBoundsException: Index: 6, Size: 4
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at tokenJar.DataModel.getByName(DataModel.java:136)
	at burp.BurpExtender.processRequestMessage(BurpExtender.java:184)
	at burp.BurpExtender.processHttpMessage(BurpExtender.java:78)
	at burp.vec.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
java.lang.IndexOutOfBoundsException: Index: 6, Size: 4
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at tokenJar.DataModel.getByName(DataModel.java:136)
	at burp.BurpExtender.processRequestMessage(BurpExtender.java:184)
	at burp.BurpExtender.processHttpMessage(BurpExtender.java:78)
	at burp.vec.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
java.lang.IndexOutOfBoundsException: Index: 6, Size: 4
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at tokenJar.DataModel.getByName(DataModel.java:136)
	at burp.BurpExtender.processRequestMessage(BurpExtender.java:184)
	at burp.BurpExtender.processHttpMessage(BurpExtender.java:78)
	at burp.vec.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
java.lang.IndexOutOfBoundsException: Index: 6, Size: 4
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at tokenJar.DataModel.getByName(DataModel.java:136)
	at burp.BurpExtender.processRequestMessage(BurpExtender.java:184)
	at burp.BurpExtender.processHttpMessage(BurpExtender.java:78)
	at burp.vec.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
java.lang.IndexOutOfBoundsException: Index: 6, Size: 4
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at tokenJar.DataModel.getByName(DataModel.java:136)
	at burp.BurpExtender.processRequestMessage(BurpExtender.java:184)
	at burp.BurpExtender.processHttpMessage(BurpExtender.java:78)
	at burp.vec.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
java.lang.IndexOutOfBoundsException: Index: 6, Size: 4
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at tokenJar.DataModel.getByName(DataModel.java:136)
	at burp.BurpExtender.processRequestMessage(BurpExtender.java:184)
	at burp.BurpExtender.processHttpMessage(BurpExtender.java:78)
	at burp.vec.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
java.lang.IndexOutOfBoundsException: Index: 6, Size: 4
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at tokenJar.DataModel.getByName(DataModel.java:136)
	at burp.BurpExtender.processRequestMessage(BurpExtender.java:184)
	at burp.BurpExtender.processHttpMessage(BurpExtender.java:78)
	at burp.vec.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
java.lang.IndexOutOfBoundsException: Index: 6, Size: 4
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at tokenJar.DataModel.getByName(DataModel.java:136)
	at burp.BurpExtender.processRequestMessage(BurpExtender.java:184)
	at burp.BurpExtender.processHttpMessage(BurpExtender.java:78)
	at burp.vec.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)

```

**Comment by DanNegrea** [comment]
Hi Andrej,

Can you please provide more info about your setup. A screenshot would help here. 

Thanks,
Dan

**Comment by Hipapheralkus** [comment]
Certainly, here is the setup with redacted parameter names. Note that I tried regexp and it matched there, so it shouldn't be a problem, but I can be mistaken. Or you had some other screenshot on mind? Not that the TokenJar was enabled when I got exceptions, so I just loaded it disabled for the screenshot.
![tokenjar](https://user-images.githubusercontent.com/4717664/38974674-4ccc9b72-43ab-11e8-87f0-5372c95a4ee2.png)

**Comment by DanNegrea** [comment]
You are a core user 👍 

I will try to fix this problem, but first I need to understand what triggers it.

Do you know for which of these params the exception is generated? I am more interested in what type of parameter it is. Body for example might not be the right answer because XML or JSON are in the body but have their own type (see [here](https://portswigger.net/burp/extender/api/burp/IParameter.html))

Meanwhile you can try to use my [second extension](https://github.com/DanNegrea/PyRules) as a workaround (maybe just for that parameter)
This extension is more flexible but less user friendly. I have two examples that you can use.

**Comment by Hipapheralkus** [comment]
I would love to help you, but unfortunately I'm still unable to find out on which request/response it's failing. Is it possible to amend the extender to print the URL if exception is raised? It would be of great help for debugging.
It still happens, but in my scope I have too many requests and I can't pinpoint which of them throws these exceptions... I just know it happens during the Active Scan.

**Comment by DanNegrea** [comment]
Hi Andrej,

I believe I pinpointed the bug. I attached here an intermediary version. If you have a urgent need try it:

1.  make a backup of your configuration

2. download the zip file attached to this comment

3. change the extension from .zip to .jar

4. load it in Extender via the Add button

In case something goes wrong let me know. If it works send me a message.

**Note**: If the parameter is not updated, check the "other" box and try again with the name in lowercase. (This is the case for XML attributes, JSON parameters and multi part attribute)

[TokenJar.zip](https://github.com/DanNegrea/TokenJar/files/2078194/TokenJar.zip)

**Comment by Hipapheralkus** [comment]
Hi, So I now have a very large trace of the failure:) The request query contains some client information which I would rather not share here. Is it possible to send the trace to you directly in some way?

**Comment by DanNegrea** [comment]
Please send me an email: dan@sec3ty.com (PGP key available).

**Comment by Hipapheralkus** [comment]
Thank you, I believe this is fixed now, I can no longer see the Exceptions, and even for JSON content type, the values are parsed properly:)

**Comment by DanNegrea** [comment]
Closed with release 2.0
